sys/kern/vm_physmem.c

```diff
@@ -77,27 +80,34 @@ static void *vm_boot_alloc(size_t n) {
 
   assert(seg != NULL);
 
-  for (void *va = align(begin, PAGESIZE); va < end; va += PAGESIZE) {
+  for (void *va = begin; va < end; va += PAGESIZE) {
     paddr_t pa = seg->start;
     seg->start += PAGESIZE;
     if (--seg->npages == 0) {
+      assert(seg->start == seg->end);
       TAILQ_REMOVE(&seglist, seg, seglink);
-      seg = TAILQ_FIRST(&seglist);
+      /* The segment should service the whole request. */
+      seg = NULL;
     }
 
     pmap_kenter((vaddr_t)va, pa, VM_PROT_READ | VM_PROT_WRITE, 0);
   }
 
-  vm_kernel_end += n;
+  vm_kernel_end = (vaddr_t)end;
 
   return begin;
 }
 
 static void vm_boot_finish(void) {
-  vm_kernel_end = (vaddr_t)align((void *)vm_kernel_end, PAGESIZE);
   vm_boot_done = true;
 }
 
+static inline size_t pg_compute_size(paddr_t pa) {
+  /* Take the highest number of the form 2^k * PAGESIZE
```

Comment:
Do I understand correctly that you don't explain here what is `PM_NQUEUES - 1`?
(only second argument of `min`)